File: src/modules/configuration/configuration.controller.ts
import {
  Controller,
  Get,
  Put,
  Body,
  Param,
  Delete,
  UseGuards,
  Post,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth } from '@nestjs/swagger';
import { ConfigurationService } from './configuration.service';
import { UpdateConfigurationDto } from './dto/update-configuration.dto';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
import { RolesGuard } from '../auth/guards/roles.guard';
import { Roles } from '../auth/decorators/roles.decorator';

@ApiTags('Configuración')
@Controller('configuration')
@UseGuards(JwtAuthGuard, RolesGuard)
@ApiBearerAuth()
export class ConfigurationController {
  constructor(private readonly configurationService: ConfigurationService) {}

  @Get()
  @ApiOperation({ summary: 'Obtener todas las configuraciones' })
  @ApiResponse({ status: 200, description: 'Lista de configuraciones' })
  @Roles('admin')
  async findAll() {
    return this.configurationService.findAll();
  }

  @Get('system')
  @ApiOperation({ summary: 'Obtener configuración del sistema organizada por categorías' })
  @ApiResponse({ status: 200, description: 'Configuración del sistema' })
  @Roles('admin')
  async getSystemConfig() {
    return this.configurationService.getSystemConfig();
  }

  @Get('categories')
  @ApiOperation({ summary: 'Obtener categorías de configuración' })
  @ApiResponse({ status: 200, description: 'Lista de categorías' })
  @Roles('admin')
  async getCategories() {
    return this.configurationService.getCategories();
  }

  @Get('category/:category')
  @ApiOperation({ summary: 'Obtener configuraciones por categoría' })
  @ApiResponse({ status: 200, description: 'Configuraciones de la categoría' })
  @Roles('admin')
  async findByCategory(@Param('category') category: string) {
    return this.configurationService.findByCategory(category);
  }

  @Get(':key')
  @ApiOperation({ summary: 'Obtener configuración específica por clave' })
  @ApiResponse({ status: 200, description: 'Configuración encontrada' })
  @ApiResponse({ status: 404, description: 'Configuración no encontrada' })
  @Roles('admin')
  async findOne(@Param('key') key: string) {
    return this.configurationService.findOne(key);
  }

  @Get(':key/value')
  @ApiOperation({ summary: 'Obtener valor de configuración específica' })
  @ApiResponse({ status: 200, description: 'Valor de la configuración' })
  @ApiResponse({ status: 404, description: 'Configuración no encontrada' })
  @Roles('admin')
  async getValue(@Param('key') key: string) {
    const value = await this.configurationService.getValue(key);
    return { key, value };
  }

  @Put(':key')
  @ApiOperation({ summary: 'Actualizar configuración específica' })
  @ApiResponse({ status: 200, description: 'Configuración actualizada exitosamente' })
  @ApiResponse({ status: 404, description: 'Configuración no encontrada' })
  @Roles('admin')
  async update(
    @Param('key') key: string,
    @Body() updateConfigurationDto: UpdateConfigurationDto,
  ) {
    return this.configurationService.update(key, updateConfigurationDto);
  }

  @Post(':key')
  @ApiOperation({ summary: 'Establecer valor de configuración' })
  @ApiResponse({ status: 201, description: 'Configuración establecida exitosamente' })
  @Roles('admin')
  async setValue(
    @Param('key') key: string,
    @Body() body: { value: any; type?: string },
  ) {
    return this.configurationService.setValue(key, body.value, body.type);
  }

  @Delete(':key')
  @ApiOperation({ summary: 'Eliminar configuración específica' })
  @ApiResponse({ status: 200, description: 'Configuración eliminada exitosamente' })
  @ApiResponse({ status: 404, description: 'Configuración no encontrada' })
  @Roles('admin')
  async remove(@Param('key') key: string) {
    await this.configurationService.remove(key);
    return { message: `Configuración '${key}' eliminada exitosamente` };
  }

  @Post('initialize')
  @ApiOperation({ summary: 'Inicializar configuraciones por defecto' })
  @ApiResponse({ status: 201, description: 'Configuraciones inicializadas exitosamente' })
  @Roles('admin')
  async initializeDefaultConfigs() {
    await this.configurationService.initializeDefaultConfigs();
    return { message: 'Configuraciones por defecto inicializadas exitosamente' };
  }

  @Post('cache/clear')
  @ApiOperation({ summary: 'Limpiar cache de configuraciones' })
  @ApiResponse({ status: 200, description: 'Cache limpiado exitosamente' })
  @Roles('admin')
  async clearCache() {
    this.configurationService.clearCache();
    return { message: 'Cache de configuraciones limpiado exitosamente' };
  }
} 

File: src/main/java/com/yego/backend/controller/yego-principal/ConfigurationController.java
package com.yego.backend.controller.yego_principal;

import com.yego.backend.entity.yego_principal.api.*;
import com.yego.backend.service.yego_principal.ConfigurationService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import jakarta.persistence.EntityNotFoundException;
import jakarta.validation.Valid;
import java.util.List;
import java.util.Map;

/**
 * Controlador REST para configuraciones del sistema YEGO Principal
 * Equivalente a ConfigurationController de NestJS
 */
@Slf4j
@RestController
@RequestMapping("/api/yego-principal/configuration")
@RequiredArgsConstructor
public class ConfigurationController {
    
    private final ConfigurationService configurationService;
    
    /**
     * Obtener todas las configuraciones
     */
    @GetMapping
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> findAll() {
        try {
            List<ConfigurationResponseDto> configurations = configurationService.findAll();
            return ResponseEntity.ok(configurations);
            
        } catch (Exception e) {
            log.error("Error obteniendo configuraciones YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener configuración del sistema organizada por categorías
     */
    @GetMapping("/system")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getSystemConfig() {
        try {
            SystemConfigDto systemConfig = configurationService.getSystemConfig();
            return ResponseEntity.ok(systemConfig);
            
        } catch (Exception e) {
            log.error("Error obteniendo configuración del sistema YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener categorías de configuración
     */
    @GetMapping("/categories")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getCategories() {
        try {
            List<String> categories = configurationService.getCategories();
            return ResponseEntity.ok(categories);
            
        } catch (Exception e) {
            log.error("Error obteniendo categorías YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener configuraciones por categoría
     */
    @GetMapping("/category/{category}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> findByCategory(@PathVariable String category) {
        try {
            List<ConfigurationResponseDto> configurations = configurationService.findByCategory(category);
            return ResponseEntity.ok(configurations);
            
        } catch (Exception e) {
            log.error("Error obteniendo configuraciones por categoría YEGO Principal {}: {}", category, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener configuración específica por clave
     */
    @GetMapping("/{key}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> findOne(@PathVariable String key) {
        try {
            ConfigurationResponseDto configuration = configurationService.findOne(key);
            return ResponseEntity.ok(configuration);
            
        } catch (EntityNotFoundException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("message", e.getMessage()));
        } catch (Exception e) {
            log.error("Error obteniendo configuración YEGO Principal {}: {}", key, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener valor de configuración específica
     */
    @GetMapping("/{key}/value")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getValue(@PathVariable String key) {
        try {
            Object value = configurationService.getValue(key);
            ConfigurationValueDto response = ConfigurationValueDto.builder()
                    .key(key)
                    .value(value)
                    .build();
            
            return ResponseEntity.ok(response);
            
        } catch (EntityNotFoundException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("message", e.getMessage()));
        } catch (Exception e) {
            log.error("Error obteniendo valor de configuración YEGO Principal {}: {}", key, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Actualizar configuración específica
     */
    @PutMapping("/{key}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> update(@PathVariable String key, 
                                   @Valid @RequestBody UpdateConfigurationDto updateConfigurationDto) {
        try {
            ConfigurationResponseDto configuration = configurationService.update(key, updateConfigurationDto);
            
            log.info("⚙️ Configuración YEGO Principal actualizada: {}", key);
            
            return ResponseEntity.ok(configuration);
            
        } catch (EntityNotFoundException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("message", e.getMessage()));
        } catch (Exception e) {
            log.error("Error actualizando configuración YEGO Principal {}: {}", key, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Establecer valor de configuración
     */
    @PostMapping("/{key}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> setValue(@PathVariable String key, 
                                     @Valid @RequestBody SetValueDto setValueDto) {
        try {
            ConfigurationResponseDto configuration = configurationService.setValue(
                    key, setValueDto.getValue(), setValueDto.getType());
            
            log.info("⚙️ Valor de configuración YEGO Principal establecido: {} = {}", key, setValueDto.getValue());
            
            return ResponseEntity.status(HttpStatus.CREATED).body(configuration);
            
        } catch (Exception e) {
            log.error("Error estableciendo valor de configuración YEGO Principal {}: {}", key, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Eliminar configuración específica
     */
    @DeleteMapping("/{key}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> remove(@PathVariable String key) {
        try {
            configurationService.remove(key);
            
            log.info("🗑️ Configuración YEGO Principal eliminada: {}", key);
            
            return ResponseEntity.ok(Map.of("message", "Configuración '" + key + "' eliminada exitosamente"));
            
        } catch (EntityNotFoundException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("message", e.getMessage()));
        } catch (Exception e) {
            log.error("Error eliminando configuración YEGO Principal {}: {}", key, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Inicializar configuraciones por defecto
     */
    @PostMapping("/initialize")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> initializeDefaultConfigs() {
        try {
            configurationService.initializeDefaultConfigs();
            
            log.info("⚙️ Configuraciones por defecto YEGO Principal inicializadas");
            
            return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
                "message", "Configuraciones por defecto inicializadas exitosamente"
            ));
            
        } catch (Exception e) {
            log.error("Error inicializando configuraciones por defecto YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Limpiar cache de configuraciones
     */
    @PostMapping("/cache/clear")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> clearCache() {
        try {
            configurationService.clearCache();
            
            log.info("🧹 Cache de configuraciones YEGO Principal limpiado");
            
            return ResponseEntity.ok(Map.of("message", "Cache de configuraciones limpiado exitosamente"));
            
        } catch (Exception e) {
            log.error("Error limpiando cache de configuraciones YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
}
